gms/gossiper.cc

```diff
@@ -2028,28 +2070,108 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
     });
 }
 
+bool operator==(const gossip_query_token_status_response& x, gossip_query_token_status_response& y) {
+    return x.nr_known_nodes == y.nr_known_nodes
+            && x.normal_token_hash == y.normal_token_hash
+            && x.bootstrap_token_hash == y.bootstrap_token_hash
+            && x.leaving_nodes_hash == y.leaving_nodes_hash
+            && x.status_hash == y.status_hash;
+}
+
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with gossip query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
-    if (force_after == 0) {
-        return do_enable_features();
-    }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
+    return wait_for_gossip_with_query().then([do_enable_features] {
```

Comment:
On restart, all known peer nodes are marked as DOWN by default. Marking of a node up is triggered when major gossip application update is received. Once wait_for_gossip_with_query is done, we can guarantee this restarting node has received application updates from all the peers, which in turn should have marked the node as UP.  The mark UP operation calls the GOSSIP_ECHO verb to really mark a node up. wait_for_gossip_with_query can not guarantee the GOSSIP_ECHO will succeed, after which the restarting node mark a peer as UP. However, it does not matter because if GOSSIP_ECHO fails due to fail to talk to a peer, the data query verb will fail as well.

Reply:
The problem is with the scenario in which GOSSIP_ECHO will not fail, so the node would be eventually marked as UP, but we're not waiting for this, and allow CQL requests to execute before the nodes are marked as UP.

Reply:
We do not wait for the nodes to be marked as UP even before this patch. 

It is wrong to require all the nodes to be UP before gossiper::wait_for_gossip_with_query returns. 

For example, 

1) Wait for the checksum to match
2) Wait for the nodes to be marked as UP

Between 1) and 2), a node can become DOWN so 2 will never return. It is totally OK for a node to become DOWN.

Note when CASSANDRA-4288 is reported there is even no waiting for gossip to settle in the code. In practice after wait_for_gossip_with_query all nodes that are really UP should  be marked as UP.

Perhaps we could do this:

1) Wait for the checksum to match
2) Check if all nodes are marked as UP. if yes, go ahead, if no, wait for _pending_mark_alive_endpoints to become empty.

Reply:
I did not suggest that we should wait for all the nodes to be UP. We should wait for nodes which are actually up to be marked as UP. That was the purpose of this wait here.

Waiting for pending mark-alives should do it, but I wouldn't use busy polling to implement this but rather a `phased_barrier`.

Reply:
> I did not suggest that we should wait for all the nodes to be UP. We should wait for nodes which are actually up to be marked as UP. That was the purpose of this wait here.

I know you mean waiting for the nodes that are alive. The example I gave above shows such wait is problematic. However, waiting _pending_mark_alive_endpoints is fine. Because, it does not match what is the finally status of a node (up or down), we just need to make sure we at least marked it once. 

```
 1  Wait for the checksum to match
  2  Wait for the nodes to be marked as UP

Between 1) and 2), a node can become DOWN so 2 will never return. It is totally OK for a node to become DOWN.
```
 
> 
> Waiting for pending mark-alives should do it, but I wouldn't use busy polling to implement this but rather a `phased_barrier`.

Good.